### PR TITLE
Add channel to pass pRnd from DRG to consensus; Make consensus engine a separate package to break import cycle.

### PR DESCRIPTION
**Pull request by rlan35**
2 shards, 78 consensus, 1312 total TPS, 11 nodes 127.0.0.1, 1, 719.096

### CODE REVIEW

File: core/blockchain.go [hunk 1]
```diff
@@ -223,6 +223,22 @@ func (bc *BlockChain) ValidateNewBlock(block *types.Block, address common.Addres
 	return nil
 }
 
+// IsEpochBlock returns whether this block is the first block of an epoch.
```
Comment (chaosma) [inline core/blockchain.go]: I have a similar function  CheckEpochBlock in core/resharding.go. So we don't need to duplicate it

File: core/blockchain.go [hunk 2]
```diff
@@ -223,6 +223,22 @@ func (bc *BlockChain) ValidateNewBlock(block *types.Block, address common.Addres
...
+func IsEpochBlock(block *types.Block) bool {
+	if block.NumberU64()%BlocksPerEpoch == 0 {
+		return true
+	}
+	return false
+}
+
+// IsEpochLastBlock returns whether this block is the last block of an epoch.
```
Comment (chaosma) [inline core/blockchain.go]: maybe move this to core/resharding.go ?

File: drand/drand.go [hunk 3]
```diff
@@ -23,7 +23,8 @@ type DRand struct {
 	bitmap                *bls_cosi.Mask
 	pRand                 *[32]byte
 	rand                  *[32]byte
```
Comment (chaosma) [inline drand/drand.go]: are we going to change the number of bytes in future? If so, make it as a variable

File: core/chain_makers.go [hunk 4]
```diff
@@ -251,7 +251,7 @@ func makeHeaderChain(parent *types.Header, n int, engine consensus.Engine, db et
 }
 
 // makeBlockChain creates a deterministic chain of blocks rooted at parent.
-func makeBlockChain(parent *types.Block, n int, engine consensus.Engine, db ethdb.Database, seed int) []*types.Block {
+func makeBlockChain(parent *types.Block, n int, engine consensus_engine.Engine, db ethdb.Database, seed int) []*types.Block {
 	blocks, _ := GenerateChain(params.TestChainConfig, parent, engine, db, n, func(i int, b *BlockGen) {
 		b.SetCoinbase(common.Address{0: byte(seed), 19: byte(i)})
```
Comment (chaosma) [inline core/chain_makers.go]: what's the meaning of 0 and 19 here?

File: drand/drand.go (hunk 3 above)
Comment (rlan35) [reply drand/drand.go]: I will change it to the correct size after adding real vrf.

File: core/chain_makers.go (hunk 4 above)
Comment (rlan35) [reply core/chain_makers.go]: This is ethereum code. It's probably some default address.

File: core/blockchain.go (hunk 1 above)
Comment (rlan35) [reply core/blockchain.go]: I think the checking is better to be on blockchain level. So I switched the usages of yours to mine.

File: core/blockchain.go (hunk 2 above)
Comment (rlan35) [reply core/blockchain.go]: I kept both here and switched yours to mine.

**Pull request merged**